Fix #4739 Talkback does not read the expected text

Vishwajith-Shettigar opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
Fix #4739, edittext requesting focus only when reader is off will resolve this issue.
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
Before fix

https://github.com/oppia/oppia-android/assets/76042077/41bc96b4-3fe1-4688-a77b-02bfdbd22782


After fix

  Talkback On

https://github.com/oppia/oppia-android/assets/76042077/a01da3dc-8848-4632-a8c0-17ce3e409beb


Talkback Off


https://github.com/oppia/oppia-android/assets/76042077/beac309e-16d8-496f-bc09-af8de634536b





 

<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

Vishwajith-Shettigar committed:
talkbackissuefixed

Vishwajith-Shettigar committed:
talkbackissuefixed

Vishwajith-Shettigar commented:
@adhiamboperes PTAL

adhiamboperes reviewed (changes requested):
Thanks @Vishwajith-Shettigar. Please add a talkback screen recording to your PR description. You also need to add tests as mentioned in the issue discussion.

Vishwajith-Shettigar committed:
talkbackissuefixed

Vishwajith-Shettigar commented:
> Thanks @Vishwajith-Shettigar. Please add a talkback screen recording to your PR description. You also need to add tests as mentioned in the issue discussion.

@adhiamboperes I attached screen recordings

Vishwajith-Shettigar commented:
> > Thanks @Vishwajith-Shettigar. Please add a talkback screen recording to your PR description. You also need to add tests as mentioned in the issue discussion.
> 
> @adhiamboperes,PTAL I have attached screen recordings. Thank you

Vishwajith-Shettigar committed:
talkbackissuefixedd

Vishwajith-Shettigar committed:
talkbackissuefixedd

Vishwajith-Shettigar committed:
talkbackissuefixedd

Vishwajith-Shettigar committed:
talkbackissuefixedtest

adhiamboperes commented on app/src/main/java/org/oppia/android/app/profile/PinPasswordActivityPresenter.kt:
```diff
@@ -55,6 +55,14 @@ class PinPasswordActivityPresenter @Inject constructor(
       (activity as PinPasswordActivity).finish()
     }
 
+    binding.apply {
+      pinPasswordInputPinEditText.setOnClickListener {
+        pinPasswordInputPinEditText.isFocusable = true
+        pinPasswordInputPinEditText.isFocusableInTouchMode = true
+        pinPasswordInputPinEditText.requestFocus()
+      }
+    }
```
```suggestion
    binding.pinPasswordInputPinEditText.setOnClickListener { pinEditText ->
        pinEditText.isFocusable = true
        pinEditText.requestFocus()
      }
```

You have already set pinEditText.isFocusableInTouchMode = true in xml.

Vishwajith-Shettigar replied:
Got it.Thank you

adhiamboperes reviewed (changes requested):
Thanks @Vishwajith-Shettigar.

So the "after" video you added does not fully illustrate that the problem has been solved. Instead of interrupting talkback, could you please let it read out the full text? I expect all the text on the screen to be read until the edittext. If you find that the textviews are not being read then it means that they have no content description, and you shoukd add some.

Vishwajith-Shettigar commented:
> Thanks @Vishwajith-Shettigar.
> 
> So the "after" video you added does not fully illustrate that the problem has been solved. Instead of interrupting talkback, could you please let it read out the full text? I expect all the text on the screen to be read until the edittext. If you find that the textviews are not being read then it means that they have no content description, and you shoukd add some.

I thought textviews text should be read only when user click on them . So you are saying when we enter activity, talk back should read all the text in the textviews automatically until edit text, am i right ? please let me know if there is any issue with what i'm actually thinking.

adhiamboperes commented:
> > Thanks @Vishwajith-Shettigar.
> > So the "after" video you added does not fully illustrate that the problem has been solved. Instead of interrupting talkback, could you please let it read out the full text? I expect all the text on the screen to be read until the edittext. If you find that the textviews are not being read then it means that they have no content description, and you shoukd add some.
> 
> I thought textviews text should be read only when user click on them . So you are saying when we enter activity, talk back should read all the text in the textviews automatically until edit text, am i right ? please let me know if there is any issue with what i'm actually thinking.


Technically yes, the swipe gestures that gets them in focus. In your video demo, I did not see that.

Vishwajith-Shettigar commented:
> > > Thanks @Vishwajith-Shettigar.
> > > So the "after" video you added does not fully illustrate that the problem has been solved. Instead of interrupting talkback, could you please let it read out the full text? I expect all the text on the screen to be read until the edittext. If you find that the textviews are not being read then it means that they have no content description, and you shoukd add some.
> > 
> > 
> > I thought textviews text should be read only when user click on them . So you are saying when we enter activity, talk back should read all the text in the textviews automatically until edit text, am i right ? please let me know if there is any issue with what i'm actually thinking.
> 
> Technically yes, the swipe gestures that gets them in focus. In your video demo, I did not see that.

Im actually not getting, are you saying talkback should read both textview without any user gesture or click ?

Vishwajith-Shettigar commented:
> > > Thanks @Vishwajith-Shettigar.
> > > So the "after" video you added does not fully illustrate that the problem has been solved. Instead of interrupting talkback, could you please let it read out the full text? I expect all the text on the screen to be read until the edittext. If you find that the textviews are not being read then it means that they have no content description, and you shoukd add some.
> > 
> > 
> > I thought textviews text should be read only when user click on them . So you are saying when we enter activity, talk back should read all the text in the textviews automatically until edit text, am i right ? please let me know if there is any issue with what i'm actually thinking.
> 
> Technically yes, the swipe gestures that gets them in focus. In your video demo, I did not see that.

Okay  i got it, basically talk back should read toolbar close button first automatically without user interaction, and when user click or use gesture on textview then only textview should be read.

Vishwajith-Shettigar commented:
@adhiamboperes  I have updated screen recording, if there is any issue please let me know. Thank you

Vishwajith-Shettigar committed:
talkbackisssuecommit8

adhiamboperes commented on app/src/main/java/org/oppia/android/app/profile/PinPasswordActivityPresenter.kt:
```diff
@@ -55,6 +55,13 @@ class PinPasswordActivityPresenter @Inject constructor(
       (activity as PinPasswordActivity).finish()
     }
 
+    binding.apply {
+      pinPasswordInputPinEditText.setOnClickListener {
+        pinPasswordInputPinEditText.isFocusable = true
+        pinPasswordInputPinEditText.requestFocus()
+      }
+    }
+
```
```suggestion
     binding.pinPasswordInputPinEditText.setOnClickListener {  pinEditText ->
      pinEditText.isFocusable = true
      pinEditText.requestFocus()
    }

```

adhiamboperes reviewed (changes requested):
Thanks @Vishwajith-Shettigar.

The video looks good. I have only left a minor suggestion on how to improve the click listener block of code. You do not need the `.apply`. Could you PTAL at how the other bindings have been implemented in the same class?

Vishwajith-Shettigar committed:
talkbackisssuecommit9

Vishwajith-Shettigar commented:
@adhiamboperes PTAL

adhiamboperes reviewed (changes requested):
Could you please also check whether we still need this on line 72 `binding.pinPasswordInputPinEditText.requestFocus()`

Vishwajith-Shettigar commented:
> Could you please also check whether we still need this on line 72 `binding.pinPasswordInputPinEditText.requestFocus()`

okay

Vishwajith-Shettigar committed:
Merge branch 'develop' into talkbackissue

Vishwajith-Shettigar committed:
talkbackisssuecommit10

Vishwajith-Shettigar committed:
talkbackisssuecommit10

Vishwajith-Shettigar commented:
> Could you please also check whether we still need this on line 72 `binding.pinPasswordInputPinEditText.requestFocus()`

@adhiamboperes 
1.requestFocus() is needed as i seen, if i run app without requestFocus(), edit text was not getting focus at first tap, user has to tap two times. 
2. isFocusableInTouchMode=true also required  inorder to get focus and i removed it from xml layout file.
3.I added the below code because the keyboard didn't appear on the first click, so I had to include it in order to make keyboard appear on first click.

 ```
val imm = activity.applicationContext.getSystemService(Context.INPUT_METHOD_SERVICE)
        as InputMethodManager
      imm.showSoftInput(pinEditText, InputMethodManager.SHOW_IMPLICIT)
```

Thank you.

Vishwajith-Shettigar committed:
talkbackisssuecommit12

Vishwajith-Shettigar committed:
talkbackisssuecommit12

Vishwajith-Shettigar commented:
@adhiamboperes PTAL

adhiamboperes reviewed (commented):
Hey @Vishwajith-Shettigar, we are discussing making this fix in a way that does not actually change the existing behaviour of keyboard focus, and @BenHenning suggested using the `android:accessibilityTraversalBefore` and `android:accessibilityTraversalAfter` attributes. Could you please check if talkback works as expected with them(probably check out the develop branch for this).
I found this [SO link ](https://stackoverflow.com/questions/58100682/change-reading-order-of-elements-within-a-focusable-viewgroup) although it does not sound like the OP resolved this. See also [medium](https://medium.com/@bradley.wilson/android-accessibility-development-doesnt-have-to-be-scary-971cfe713a0e).

Vishwajith-Shettigar committed:
Merge branch 'oppia:develop' into talkbackissue

Vishwajith-Shettigar commented:
> Hey @Vishwajith-Shettigar, we are discussing making this fix in a way that does not actually change the existing behaviour of keyboard focus, and @BenHenning suggested using the `android:accessibilityTraversalBefore` and `android:accessibilityTraversalAfter` attributes. Could you please check if talkback works as expected with them(probably check out the develop branch for this). I found this [SO link ](https://stackoverflow.com/questions/58100682/change-reading-order-of-elements-within-a-focusable-viewgroup) although it does not sound like the OP resolved this. See also [medium](https://medium.com/@bradley.wilson/android-accessibility-development-doesnt-have-to-be-scary-971cfe713a0e).

i tried, but it didn't work, edittext gets initial focus, is there any way to avoid that ?
Even if i make edittext focusable false and making it true dynamically there are few complications that we already discussed.

Vishwajith-Shettigar commented:
@adhiamboperes  below code makes talkback work properly, but I'm not sure whether  using delay is good practice or not in this case.
` GlobalScope.launch(Dispatchers.Default) {
      delay(1000)
      binding.pinPasswordInputPinEditText.importantForAccessibility=1
    }
`

Vishwajith-Shettigar committed:
fixed talkback directly focusing edittext

Vishwajith-Shettigar committed:
Merge branch 'oppia:develop' into talkbackissue

Vishwajith-Shettigar committed:
fixed talkback directly focusing edittext

Vishwajith-Shettigar committed:
fixed talkback directly focusing edittext

Vishwajith-Shettigar committed:
fixed talkback directly focusing edittext

Vishwajith-Shettigar committed:
Merge branch 'talkbackissue' of https://github.com/Vishwajith-Shettigar/oppia-android into talkbackissue

Vishwajith-Shettigar committed:
fixed talkback directly focusing edittext

Vishwajith-Shettigar committed:
fixed talkback directly focusing edittext

Vishwajith-Shettigar committed:
fixed talkback directly focusing edittext

Vishwajith-Shettigar committed:
fixed talkback directly focusing edittext

Vishwajith-Shettigar committed:
Merge branch 'develop' into talkbackissue

Vishwajith-Shettigar committed:
Merge branch 'oppia:develop' into talkbackissue

Vishwajith-Shettigar committed:
fixing talkback issue

Vishwajith-Shettigar committed:
fixing talkback issue

Vishwajith-Shettigar committed:
talkback issue fix

Vishwajith-Shettigar committed:
talkback issue fix

adhiamboperes commented on app/src/main/java/org/oppia/android/app/profile/PinPasswordActivityPresenter.kt:
```diff
@@ -69,7 +71,13 @@ class PinPasswordActivityPresenter @Inject constructor(
     }
-    binding.pinPasswordInputPinEditText.requestFocus()
+
+    val am = (activity).getSystemService(Context.ACCESSIBILITY_SERVICE) as AccessibilityManager
+    val isTalkbackEnabled = am.isTouchExplorationEnabled()
+
+    if (!isTalkbackEnabled)
+      binding.pinPasswordInputPinEditText.requestFocus()
+
```
While this is technically correct, we have our own accessibility service that wraps this functionality and can easily be swapped out in tests. See this PR: https://github.com/oppia/oppia-android/pull/5027/files

```suggestion
if (!accessibilityService.isScreenReaderEnabled()) {
binding.pinPasswordInputPinEditText.requestFocus()
}
```

adhiamboperes replied:
Also per group chat discussion, make sure you add a comment explaining this functionality and why we need it.

adhiamboperes reviewed (commented):
Once you have make the suggested change, please add a test for the edittext being focussed/unfocussed when screen reader is on/off. See TopicFragmentTest in [this PR](https://github.com/oppia/oppia-android/pull/5027/files)  for reference:

Vishwajith-Shettigar committed:
fixed talkback issue

Vishwajith-Shettigar committed:
fixed talkback issue

Vishwajith-Shettigar committed:
added tests

Vishwajith-Shettigar committed:
added tests

Vishwajith-Shettigar committed:
added comment

Vishwajith-Shettigar committed:
added comment

adhiamboperes commented on app/src/main/java/org/oppia/android/app/profile/PinPasswordActivityPresenter.kt:
```diff
@@ -69,7 +71,17 @@ class PinPasswordActivityPresenter @Inject constructor(
     }
-    binding.pinPasswordInputPinEditText.requestFocus()
+
+    /**
+     * If the screen reader is off, the EditText will receive focus.
+     * If the screen reader is on, the EditText won't receive focus.
+     * We need this because if we give focus to the EditText when the screen reader is on,
+     TalkBack will automatically focus on the EditText when the user enters the activity.
+     */
```
This can be a regular comment and not a KDoc:
```
//...
//...
//...
```

adhiamboperes replied:
```suggestion
     // This is needed because requesting focus on the EditText when the screen reader is on gives TalkBack priority over other views in the screen, ignoring view hierachy.
```

adhiamboperes commented on app/src/sharedTest/java/org/oppia/android/app/profile/PinPasswordActivityTest.kt:
```diff
@@ -181,7 +185,8 @@ class PinPasswordActivityTest {
 
   @Test
-  fun testPinPassword_withAdmin_keyboardIsVisibleByDefault() {
+  fun testPinPassword_withAdmin_readOff_keyboardIsVisible() {
```
```suggestion
  fun testPinPassword_withAdmin_screenReaderOff_keyboardIsVisible() {
```

adhiamboperes commented on app/src/sharedTest/java/org/oppia/android/app/profile/PinPasswordActivityTest.kt:
```diff
@@ -193,6 +198,20 @@ class PinPasswordActivityTest {
   }
 
+  @Test
+  fun testPinPassword_withAdmin_readOn_keyboardIsNotVisible() {
```
```suggestion
  fun testPinPassword_withAdmin_screenReaderOn_keyboardIsNotVisible() {
```

adhiamboperes reviewed (changes requested):
Thanks @Vishwajith-Shettigar, just some minor corrections. See in-line.

Vishwajith-Shettigar committed:
Merge branch 'develop' into talkbackissue

Vishwajith-Shettigar committed:
Merge branch 'develop' into talkbackissue

Vishwajith-Shettigar committed:
fixed talkback issue

Vishwajith-Shettigar committed:
fixed talkback issue

Vishwajith-Shettigar commented:
@adhiamboperes, PTAL.
Thank you.

adhiamboperes reviewed (approved):
LGTM, thanks @Vishwajith-Shettigar!
